cli/src/util/io.rs

```diff
@@ -57,3 +57,41 @@ where
 
 	Ok(bytes_so_far)
 }
+
+/// Helper used when converting Future interfaces to poll-based interfaces.
+/// Stores excess data that can be reused on future polls.
+#[derive(Default)]
+pub(crate) struct ReadBuffer(Option<(Vec<u8>, usize)>);
```

Comment:
This is a copy of the snippet I wrote for the tunnels SDK, https://github.com/microsoft/dev-tunnels/blob/b20ef08c719f843a1d9ba8610c52071fae6a482f/rs/src/connections/io.rs